### DISCUSSION TITLE
Compilation error v1.1.1 in docker build: using namespace ifm3d::literals

### DISCUSSION THREAD

**Discussion by KtGunn** [ifm/ifm3d#391] [Q&A] [answered]
I am creating a docker image for the OVP800. I am using base image v1.1.1. During the image build, compilation stops with the fault: 'json' is not a member of 'ifm3d'. The code line is `ifm3d::json conf = camera->Get();`

I switched to the simple test dockerfile Lola Masson suggested in response to my previous post to see if I could recreate this fault in a smaller application. I did but now that complaint was: 'literals' is not a namespace-name. The code line was "using namespace ifm3d::literals;"

Below is the Dockerfile I use. Note that 'playground' refers to the directory under 'examples' in version **1.2.1** not 1.1.1. I have v1.2.1. installed on my development system not 1.1.1 (I hope that's not the problem!):

```Dockerfile
FROM ghcr.io/ifm/ifm3d:v1.1.1-ubuntu-arm64

ARG DEBIAN_FRONTEND=noninteractive
RUN sudo apt update && sudo apt install -y\
        build-essential \
        cmake \
        coreutils \
        libgoogle-glog0v5 \
        libgoogle-glog-dev

COPY playground /home/ifm/ifm3d_playground/.

WORKDIR /home/ifm/ifm3d_playground/build

RUN cmake .. && cmake --build .
```

Below is a screenshot of the build process and the compilation fault. Note that c++17 is specified for compilation :

![ifm_literal-fault](https://user-images.githubusercontent.com/6865198/226011105-d9304ff6-e973-4b35-97e9-787220dc9525.png)

I'm hoping you can shed light on this problem and have suggestions. Perhaps I'm using an unstable version of code?

Thank you.
-Kris

**Comment by KtGunn** [comment, marked as answer]
I suspect the problem is difference between v1.1.1 and v1.2.1. I am running v1.2.1 on my PC host--no problems--and v1.1.1 in docker container on the VPU, where I have a problem. I will run the same version on both hosts.

**Reply by KtGunn** [reply]
Problem corrected. My app running with version v1.2.1 compiles & runs both on PC host and in VPU container.

**Comment by lola-masson** [comment]
Hi Kris,

The issue is most likely from the mismatch of the v1.1.1 base image with the examples from v1.2.1. If you have a look at [the changelog](https://github.com/ifm/ifm3d/blob/main/ChangeLog.md) and [the migration guide](https://api.ifm3d.com/latest/content/migration_guide/v1_2_0.html#changes-to-ifm3d-json) for versions 1.2.x, you will see that a change was included related to json, and leading to use the `ifm3d::literals` namespace, which did not exist prior to v1.2.0. (Useful hint, if you go to api.ifm3d.com/latest, you can browse the documentation for the latest (unreleased) version of ifm3d).

Edit: somehow missed your second answer before writing this message--glad we got to the same conclusion!
